File: config/initializers/cors.rb
# Be sure to restart your server when you modify this file.

# Avoid CORS issues when API is called from the frontend app.
# Handle Cross-Origin Resource Sharing (CORS) in order to accept cross-origin Ajax requests.
#
# Origins come from CORS_ORIGINS (comma-separated). Empty means no origin is allowed.

# Read more: https://github.com/cyu/rack-cors

Rails.application.config.middleware.insert_before 0, Rack::Cors do
  allowed = ENV.fetch("CORS_ORIGINS", "").split(",").map(&:strip).reject(&:empty?)

  if allowed.any?
    allow do
      origins(*allowed)

      resource "*",
        headers: :any,
        methods: %i[get post put patch delete options head],
        expose: %w[Authorization X-Request-Id]
    end
  end
end

File: src/middleware/cors.rs
use axum::http::{HeaderName, HeaderValue, Method};
use tower_http::cors::{AllowOrigin, Any, CorsLayer};

/// Builds the CORS layer from `CORS_ORIGINS` (comma-separated). An empty list
/// means no origin is allowed.
pub fn layer() -> CorsLayer {
    let configured = std::env::var("CORS_ORIGINS").unwrap_or_default();
    let origins: Vec<HeaderValue> = configured
        .split(',')
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .filter_map(|value| HeaderValue::from_str(value).ok())
        .collect();

    CorsLayer::new()
        .allow_origin(AllowOrigin::list(origins))
        .allow_methods([
            Method::GET,
            Method::POST,
            Method::PUT,
            Method::PATCH,
            Method::DELETE,
            Method::OPTIONS,
            Method::HEAD,
        ])
        .allow_headers(Any)
        .expose_headers([
            HeaderName::from_static("authorization"),
            HeaderName::from_static("x-request-id"),
        ])
}
